packages/aws-cdk-lib/aws-s3/README.md

```diff
@@ -622,6 +622,8 @@ as it does not contain any objects.
 To override this and force all objects to get deleted during bucket deletion,
 enable the`autoDeleteObjects` option.
 
+In order to facilitate bucket deletion, we add the `s3:PutBucketPolicy` action from AWS Identity and Access Management (IAM) policies. This action authorizes users to manage bucket policies for Amazon S3 buckets, enabling them to adjust access control settings by either establishing new policies or modifying existing ones. As part of this we update the policy to **Deny** `s3:PutObject` action within the specified bucket preventing new objects from being created in the bucket. Consequently, this measure ensures that during the bucket emptying phase, the onset of potential race conditions is mitigated, as no new objects can be generated within the bucket.
```

Comment:
What I'm still not seeing is why we only added `s3:PutBucketPolicy`. As far as I can tell, there was no `Deny` for `s3:PutObject` added which is the claim in the original PR. Here is one of the snapshot diffs:
<img width="293" alt="image" src="https://github.com/aws/aws-cdk/assets/131073567/7cf729b5-c52a-47bf-9190-61eaa5dff5ad">
You can see the only difference is that `s3:PutBucketPolicy` was added with an `Allow` effect. Customers are going to be wondering why a new action was added to the bucket policy that grants more actions to be taken on the bucket. I'm even wondering why that's the case. This is confusing because they claimed that their change, "Adds a DENY policy for S3:PutObject on buckets to be auto-deleted to prevent a race condition on emptying with external bucket writers.", but that isn't at all what changed. My guess is that this was done because if you don't have `PutBucketPolicy` permissions, Amazon S3 returns a 403 Access Denied error which is essentially blocking any writes to the bucket from external bucket writers. This is what needs to be investigated a bit more and explained.

Reply:
@colifran Yes, the only difference is that `s3:PutBucketPolicy` was added with an `Allow` effect but, that policy is used to grant permissions for putting or updating bucket policies for Amazon S3 buckets. As part of this [code change](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/custom-resource-handlers/lib/aws-s3/auto-delete-objects-handler/index.ts#L46) we retrieve the existing bucket policy for the specified bucket using `getBucketPolicy` method and then add a new statement to `Deny` the `s3:PutObject` as part of this [code](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/custom-resource-handlers/lib/aws-s3/auto-delete-objects-handler/index.ts#L52). We then use the `s3:PutBucketPolicy` to update the policy while cleanup. This runs in the background while enabling the `autoDeleteObjects`. 

But, do we want to specify why that new policy is getting added and what happens if not?

Reply:
Ahh okay awesome. That part was missing for me. I would assume this would be confusing for customers too because an update for a customer using `autoDeleteObjects` will just show the `s3:PutBucketPolicy` action being added to the bucket policy in the diff. Without understanding that this is backed by a custom resource they'll likely not understand the point of that action being added. So what we want to clearly convey is that when `autoDeleteObjects` is enabled we add the `s3:PutBucketPolicy` action to the bucket policy. We do this because on bucket deletion the custom resource provider needs to update the bucket policy by adding a DENY policy for s3:PutObject to prevent a race condition on emptying with external bucket writers.